design/src/rfcs/rfc0003_presigning_api.md

```diff
@@ -0,0 +1,192 @@
+RFC: API for Pre-signed URLs
+============================
+
+> Status: RFC
+
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+Several AWS services allow for pre-signed requests in URL form, which is described well by
+[S3's documentation on authenticating requests using query parameters](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html).
+
+This doc establishes the customer-facing API for creating these pre-signed URLs and how they will
+be implemented in a generic fashion in the SDK codegen.
+
+Terminology
+-----------
+
+To differentiate between the clients that are present in the generated SDK today, the following
+terms will be used throughout this doc:
+
+- **Smithy Client**: A `smithy_client::Client<C, M, R>` struct that is responsible for gluing together
+  the connector, middleware, and retry policy. This is not generated and lives in the `smithy-client` crate.
+- **Fluent Client**: A code-generated `Client<C, M, R>` that has methods for each service operation on it.
+  A fluent builder is generated alongside it to make construction easier.
+
+Pre-signed URL config
+---------------------
+
+Today, pre-signed URLs take an expiration time that's not part of the service API.
+The SDK will make this configurable as a separate struct so that there's no chance of name collisions, and so
+that additional fields can be added in the future. Fields added later will require defaulting for
+backwards compatibility.
+
+Since there's only one configuration field today, an `expires` function will be added to conveniently
+construct `PresigningConfig` without a builder. At the same time, a builder will be present for future-proofing:
+
+```rust
+#[non_exhaustive]
+#[derive(Debug, Clone)]
+pub struct PresigningConfig {
+    expires: Duration,
```

Comment:
should we have a default expiration? Looks like the CLI defaults to 3600 seconds. We should probably also document & warn about the case where the credentials expire before the presigned URL (the request will start failing after the credentials expire)

Reply:
I think we may just need to make a judgement call here. It looks like the Ruby SDK defaults to 900 seconds for S3 GetObject, and the Java SDK doesn't have a default for the same operation. I'm kind of leaning towards not having a default since changing a default later would be a breaking change, and there's no guarantee our chosen default is the right one for the majority of use-cases.